frontend_csv/src/parser/parser.rs

```diff
@@ -0,0 +1,230 @@
+    pub fn new() -> Self {
+        return Parser {
+            pos: 0,
+            simple_table: BTreeMap::new(),
+            lnmsg_buffer: None,
+            lntlv_buffer: None,
+        };
+    }
+
+    /// Take the element in the current position of the stream
+    fn peek<'a>(&self, tokens: &'a Vec<CSVToken>) -> &'a CSVToken {
+        return &tokens[self.pos];
+    }
+
+    /// Take the element in the current position of the stream
+    /// and increase the position by one
+    fn pop(&mut self, tokens: &Vec<CSVToken>) -> CSVToken {
+        let res = tokens[self.pos].clone();
+        self.pos += 1;
+        return res;
+    }
+
+    /// Parse a message type line of the csv file, where the format looks like
+    /// the following one:
+    ///
+    /// `msgtype,init,16`
+    fn parse_msg_typ(&mut self, tokens: &Vec<CSVToken>) {
+        let msg_name = self.pop(&tokens);
+        let msg_type = self.pop(&tokens);
+        match msg_type.ty {
+            CSVTokenType::Number => {
+                self.lnmsg_buffer = Some(LNMsg {
+                    msg_typ: msg_type.val.parse::<u64>().unwrap(),
+                    msg_name: msg_name.val.to_string(),
+                    msg_data: Vec::new(),
+                    tlv_stream: Vec::new(),
+                });
+            }
+            _ => panic!("Unknown Token {:?}", self.peek(&tokens)),
+        }
+    }
+
+    fn get_byte_length(&self, keyword: String) -> u64 {
+        let tmp_vec = self.lnmsg_buffer.clone().unwrap().msg_data;
+
+        for i in 0..tmp_vec.len() {
+            match tmp_vec[i] {
+                LNMsData::Unsigned64(ref name, _) => {
+                    if name == &keyword {
+                        return 8;
+                    }
+                }
+                LNMsData::Unsigned32(ref name, _) => {
+                    if name == &keyword {
+                        return 4;
+                    }
+                }
+                LNMsData::Unsigned16(ref name, _) => {
+                    if name == &keyword {
+                        return 2;
+                    }
+                }
+                _ => {
+                    continue;
+                }
+            }
+        }
+        return 0;
+    }
+
+    //// Parse a message data entry
+    ///  msgdata,init,globalfeatures,byte,gflen
+    ///  msgdata,init,gflen,u16,
+    fn parse_msg_data(&mut self, tokens: &Vec<CSVToken>) {
+        self.pop(&tokens); // msgdata
+        let msg_data_name = self.pop(&tokens).val.to_string();
+        print!("\n{:?}\n", msg_data_name);
+        let msg_data_type = self.pop(&tokens);
+        print!("\n{:?} {:?}\n", self.pos - 1, tokens[self.pos - 1]);
+
+        print!("\n msg type {:?} \n", msg_data_type.ty);
+        match msg_data_type.ty {
+            CSVTokenType::U16 => self
+                .lnmsg_buffer
+                .as_mut()
+                .unwrap()
+                .msg_data
+                .push(LNMsData::Unsigned16(msg_data_name, 2)),
```

Comment:
```suggestion
            CSVTokenType::U16 => self
                .lnmsg_buffer
                .as_mut()
                .unwrap()
                .msg_data
                .push(LNMsData::Unsigned16(msg_data_name, 2)),
```

You can create a new method add tokens and avoid using this bad method of chaining and code duplication!

You can add a method add_data in the LNMessage structure is I'm not wrong, and simplify this as `self.lnmsg_buffer.unwrap().add_data(LNMsData::Unsigned16(msg_data_name, 2))`

And if the `lnmsg_buffer` need to be mutated, please declare it as mutable to avoid confusion

Reply:
for lnmsg_buffer shouldnt it be enough just for &mut self? as it is self.lnmsg_buffer

Reply:
I do not like the chaining here, this changing should be in a single method like `add_data`

Reply:
I change this part in here https://github.com/lnspec-tools/lncodegen.rs/tree/macros/fix_parser to return the value

Reply:
Hi I dont get why in your branch you added this in the parse_msg_data method at line 100

while self.peek(tokens).ty == CSVTokenType::MsgData

Reply:
because the message is the only type to parse and not message type and after message data is like parsing an if you have a single method that parse all the if condition, and not two method that parse the `if` and another method that parse `else`